src/java/org/apache/cassandra/utils/btree/BTree.java

```diff
@@ -3405,301 +3396,161 @@ private void clearBranchBuffer(Object[] array)
-                            unode = update.node(); upos = update.position(); uub = (Existing) update.upperBound(); usz = shallowSizeOfBranch(unode);
-                        }
-                        else if (insert.ascendToUnfinishedParent())
-                        {
-                            // if inode has been exhausted, move it forwards
-                            inode = insert.node(); ipos = insert.position(); isz = shallowSizeOfBranch(inode);
-                        }
-                        else break;
-                    }
-                    else
-                    {
-                        // {@code insert} is a branch: perform a linear merge between the leaf and this one branch key
-                        Insert ik = (Insert) inode[ipos];
-                        int find = exponentialSearchWithUpperBound(comparator, unode, upos, usz, uub, ik);
-                        boolean isInRange = find != -(2 + usz);
-                        boolean found = find >= 0;
-                        if (find < 0) find = isInRange ? -(1 + find) : usz;
-
-                        if (upos < find)
-                        {
-                            copy(unode, upos, find - upos);
-                            upos = find;
-                        }
+                final Existing nextUKey = find < usz ? (Existing) unode[find] : uub;
+                final Object[] childUNode = (Object[]) unode[find + usz];
 
-                        if (isInRange)
-                        {
-                            leaf().addKey(found ? updateF.apply((Existing) unode[upos++], ik)
-                                                : isSimple(updateF)
-                                                  ? ik
-                                                  : updateF.apply(ik));
-
-                            // We are now immediately after ik; since inode is a branch, the next is in a leaf node
-                            insert.descendIntoNextLeaf(inode, ipos, isz);
-                            inode = insert.node(); ipos = insert.position(); isz = sizeOfLeaf(inode);
-                        }
-                        else
-                        {
-                            if (null == (level = branch = finish(level, unode, usz)))
-                                break;
-                            unode = update.node(); upos = update.position(); uub = (Existing) update.upperBound(); usz = shallowSizeOfBranch(unode);
-                        }
-                    }
+                // process next child
+                if (c < 0)
+                {
+                    // ik fall inside it -- recursively merge the child with the update, using next key as an upper bound
+                    LeafOrBranchBuilder childBuilder = builder.child;
+                    ik = updateRecursive(ik, childUNode, nextUKey, childBuilder);
+                    childBuilder.drainAndPropagate(childUNode, shallowSizeOfBranch(childUNode), builder);
```

Comment:
I think `drainAndPropagate` and `finish` no longer need to accept `usz` as a parameter as it is ordinarily computed by the caller. Would be simpler to just compute it in the method body.

Reply:
Done